lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1220,6 +1269,220 @@ class CleanupFuncOps : public FuncOpPartialLoweringPattern {
+
+    if (groupDoneOps.size() <= 1)
+      return failure();
+
+    /// Create an and-tree of all calyx::GroupDoneOp's.
+    rewriter.setInsertionPointToEnd(groupDoneOps[0]->getBlock());
+    Value acc = groupDoneOps[0].src();
+    for (auto groupDoneOp : llvm::makeArrayRef(groupDoneOps).drop_front(1)) {
+      auto newAndOp = rewriter.create<comb::AndOp>(groupDoneOp.getLoc(), acc,
+                                                   groupDoneOp.src());
+      acc = newAndOp.getResult();
+    }
+
+    /// Create a group done op with the complex expression as a guard.
+    rewriter.create<calyx::GroupDoneOp>(
+        groupOp.getLoc(),
+        rewriter.create<hw::ConstantOp>(groupOp.getLoc(), APInt(1, 1)), acc);
+    for (auto groupDoneOp : groupDoneOps)
+      rewriter.eraseOp(groupDoneOp);
+
+    return success();
+  }
+};
+
+/// Returns true if 'child' is within the child tree of 'parent'.
+template <typename TParent>
+static bool isChildOp(TParent &parent, Operation *child) {
+  bool isChild = false;
+  parent.walk([&](Operation *op) {
+    if (child == op) {
+      isChild = true;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  return isChild;
+}
+
+/// Returns the last calyx::EnableOp within the child tree of 'parentSeqOp'.
+/// If no EnableOp was found (for instance, if a "par" group is present),
+/// returns nullptr.
+static Operation *getLastSeqEnableOp(calyx::SeqOp parentSeqOp) {
+  auto &lastOp = parentSeqOp.getBody()->back();
+  if (auto enableOp = dyn_cast<calyx::EnableOp>(lastOp))
+    return enableOp.getOperation();
+  else if (auto seqOp = dyn_cast<calyx::SeqOp>(lastOp))
+    return getLastSeqEnableOp(seqOp);
+  return nullptr;
+}
+
+/// Removes common tail enable operations for sequential 'then'/'else'
+/// branches inside an 'if' operation.
+///
+///   if %a with %A {           if %a with %A {
+///     seq {                     seq {
+///       ...                       ...
+///       calyx.enable @B       } else {
+///     }                         seq {
+///   } else {              ->      ...
+///     seq {                     }
+///       ...                   }
+///       calyx.enable @B       calyx.enable @B
+///     }
+///   }
+///
+struct CommonIfTailEnablePattern : mlir::OpRewritePattern<calyx::IfOp> {
+  using mlir::OpRewritePattern<calyx::IfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(calyx::IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    /// Check if there's anything in the branches; if not,
+    /// EliminateEmptyOpPattern will eliminate a potentially
+    /// empty/invalid if statement.
+    if (ifOp.thenRegion().empty() || ifOp.thenRegion().front().empty())
+      return failure();
+    if (ifOp.elseRegion().empty() || ifOp.elseRegion().front().empty())
+      return failure();
+
+    auto thenSeqOp = dyn_cast<calyx::SeqOp>(ifOp.thenRegion().front().front());
```

Comment:
Do you foresee this ever being a `ParOp`? I think the assertion is a good idea in any case.

Reply:
I imagine this kind of optimization would be a separate matcher, since we're no longer constrained by having the last `group.enable` being equal, but rather _any_ shared `group.enable` within each of the `par` blocks in the branches can be factored out.

Reply:
Makes sense. I would add a comment explicitly saying what you have here. This explanation helped me understand why you've added the assert.